pkg/genericapiserver/config.go

```diff
@@ -223,57 +258,44 @@ func NewConfig(options *options.ServerRunOptions) *Config {
 	if options.InsecurePort > 0 {
-		insecureServingInfo = &ServingInfo{
+		insecureServingInfo := &ServingInfo{
 			BindAddress: net.JoinHostPort(options.InsecureBindAddress.String(), strconv.Itoa(options.InsecurePort)),
 		}
+		c.InsecureServingInfo = insecureServingInfo
 	}
 
-	return &Config{
-		APIGroupPrefix:            options.APIGroupPrefix,
-		APIPrefix:                 options.APIPrefix,
-		CorsAllowedOriginList:     options.CorsAllowedOriginList,
-		AuditWriter:               auditWriter,
-		EnableGarbageCollection:   options.EnableGarbageCollection,
-		EnableIndex:               true,
-		EnableProfiling:           options.EnableProfiling,
-		EnableSwaggerSupport:      true,
-		EnableSwaggerUI:           options.EnableSwaggerUI,
-		EnableVersion:             true,
-		ExternalHost:              options.ExternalHost,
-		KubernetesServiceNodePort: options.KubernetesServiceNodePort,
-		MasterCount:               options.MasterCount,
-		MinRequestTimeout:         options.MinRequestTimeout,
-		SecureServingInfo:         secureServingInfo,
-		InsecureServingInfo:       insecureServingInfo,
-		PublicAddress:             options.AdvertiseAddress,
-		ReadWritePort:             options.SecurePort,
-		ServiceClusterIPRange:     &options.ServiceClusterIPRange,
-		ServiceNodePortRange:      options.ServiceNodePortRange,
-		OpenAPIDefaultResponse: spec.Response{
-			ResponseProps: spec.ResponseProps{
-				Description: "Default Response."}},
-		OpenAPIInfo: spec.Info{
-			InfoProps: spec.InfoProps{
-				Title:   "Generic API Server",
-				Version: "unversioned",
-			},
-		},
-		MaxRequestsInFlight: options.MaxRequestsInFlight,
-		LongRunningFunc:     genericfilters.BasicLongRunningRequestCheck(longRunningRE, map[string]string{"watch": "true"}),
-	}
+	c.APIGroupPrefix = options.APIGroupPrefix
+	c.APIPrefix = options.APIPrefix
+	c.CorsAllowedOriginList = options.CorsAllowedOriginList
+	c.EnableGarbageCollection = options.EnableGarbageCollection
+	c.EnableProfiling = options.EnableProfiling
+	c.EnableSwaggerUI = options.EnableSwaggerUI
+	c.ExternalHost = options.ExternalHost
+	c.KubernetesServiceNodePort = options.KubernetesServiceNodePort
+	c.MasterCount = options.MasterCount
+	c.MinRequestTimeout = options.MinRequestTimeout
+	c.PublicAddress = options.AdvertiseAddress
+	c.ServiceClusterIPRange = &options.ServiceClusterIPRange
+	c.ServiceNodePortRange = options.ServiceNodePortRange
+	c.MaxRequestsInFlight = options.MaxRequestsInFlight
+
+	return c
 }
 
 type completedConfig struct {
 	*Config
 }
 
-// Complete fills in any fields not set that are required to have valid data. It's mutating the receiver.
+// Complete fills in any fields not set that are required to have valid data and can be derived
+// from other fields.  If you're going to `ApplyOptions`, do that first.  It's mutating the receiver.
 func (c *Config) Complete() completedConfig {
-	if c.ServiceClusterIPRange == nil {
+	if c.ServiceClusterIPRange.IP == nil {
```

Comment:
Should we check both `c.ServiceClusterIPRange == nil && c.ServiceClusterIPRange.IP == nil` ?